Source/JavaScriptCore/runtime/ArrayPrototype.cpp

```diff
@@ -1511,16 +1512,36 @@ void clearElement(double& element)
     element = PNaN;
 }
 
-template<typename T>
-ALWAYS_INLINE void copyElements(T* buffer, unsigned offset, T* source, unsigned sourceSize, IndexingType sourceType)
+template<typename T, typename U>
+ALWAYS_INLINE void copyElements(T* buffer, unsigned offset, U* source, unsigned sourceSize, IndexingType sourceType)
 {
-    if (sourceType != ArrayWithUndecided) {
-        gcSafeMemcpy(buffer + offset, source, sizeof(JSValue) * sourceSize);
+    if (sourceType == ArrayWithUndecided) {
+        for (unsigned i = sourceSize; i--;)
+            clearElement<T>(buffer[i + offset]);
         return;
     }
 
-    for (unsigned i = sourceSize; i--;)
-        clearElement<T>(buffer[i + offset]);
+    if constexpr (std::is_same_v<T, U>) {
+        gcSafeMemcpy(buffer + offset, source, sizeof(JSValue) * sourceSize);
```

Comment:
Why we use `sizeof(JSValue)` instead of `sizeof(U)` for the size of source element type?

Reply:
They are always the same size to sizeof(JSValue). This function is specifically for butterfly’s buffer filling. So sizeof(JSValue) is rather better.